File: sub-agent-router-mcp-server/src/registry.ts
import fs from 'fs';
import path from 'path';
import { AgentSpec, RegistryData } from './types.js';
import { safeJsonParse, ensureDir } from './utils.js';

export class AgentRegistry {
  private filePath: string;
  private data: RegistryData;

  constructor(stateDir: string, filePath?: string) {
    ensureDir(stateDir);
    this.filePath = filePath ? path.resolve(filePath) : path.join(stateDir, 'subagents.json');
    this.data = this.load();
  }

  load(): RegistryData {
    if (!fs.existsSync(this.filePath)) {
      const initial: RegistryData = { agents: [] };
      fs.writeFileSync(this.filePath, JSON.stringify(initial, null, 2), 'utf8');
      return initial;
    }
    const raw = fs.readFileSync(this.filePath, 'utf8');
    const parsed = safeJsonParse<RegistryData>(raw, { agents: [] });
    if (!Array.isArray(parsed.agents)) {
      parsed.agents = [];
    }
    return parsed;
  }

  save(data: RegistryData) {
    this.data = data;
    fs.writeFileSync(this.filePath, JSON.stringify(this.data, null, 2), 'utf8');
  }

  getFilePath(): string {
    return this.filePath;
  }

  listAgents(): AgentSpec[] {
    return this.data.agents.slice();
  }

  getAgent(id: string): AgentSpec | null {
    return this.data.agents.find((a) => a.id === id) || null;
  }

  upsertAgent(agent: AgentSpec) {
    const index = this.data.agents.findIndex((a) => a.id === agent.id);
    if (index >= 0) {
      this.data.agents[index] = agent;
    } else {
      this.data.agents.push(agent);
    }
    this.save(this.data);
  }
}

File: sub-agent-router-mcp-server-rs/src/registry.rs
use crate::types::{AgentSpec, RegistryData};
use crate::utils::{ensure_dir, safe_json_parse};
use std::fs;
use std::path::Path;

pub struct AgentRegistry {
    data: RegistryData,
}

impl AgentRegistry {
    pub fn new(state_dir: &Path, file_path: Option<&Path>) -> Result<Self, String> {
        ensure_dir(state_dir).map_err(|err| err.to_string())?;
        let file_path = file_path
            .map(|p| p.to_path_buf())
            .unwrap_or_else(|| state_dir.join("subagents.json"));
        let data = Self::load(&file_path)?;
        Ok(Self { data })
    }

    fn load(path: &Path) -> Result<RegistryData, String> {
        if !path.exists() {
            let initial = RegistryData { agents: Vec::new() };
            let text = serde_json::to_string_pretty(&initial).map_err(|err| err.to_string())?;
            fs::write(path, text).map_err(|err| err.to_string())?;
            return Ok(initial);
        }
        let raw = fs::read_to_string(path).map_err(|err| err.to_string())?;
        let mut parsed: RegistryData = safe_json_parse(&raw, RegistryData { agents: Vec::new() });
        if parsed.agents.is_empty() {
            parsed.agents = Vec::new();
        }
        Ok(parsed)
    }

    pub fn list_agents(&self) -> Vec<AgentSpec> {
        self.data.agents.clone()
    }
}
